### PR TITLE
[Server] Implement Gradle subprojects and scaffolding for the Python server.

### PR DESCRIPTION
**Pull request by crisidev**
## Motivation and Context
See https://github.com/awslabs/smithy-rs/issues/1367.

## Description
Implement Gradle subprojects and scaffolding for the Python server.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by crisidev** [comment]
There is a failing CI action, but it is expected as I am creating a bunch of new projects and the diff tool will need them on `main` before being able to generate the diff.

File: codegen-server-test/python/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+plugins { id("software.amazon.smithy") }
+
+val smithyVersion: String by project
+val defaultRustFlags: String by project
+val defaultRustDocFlags: String by project
+val properties = PropertyRetriever(rootProject, project)
+
+val pluginName = "rust-server-codegen"
+val workingDirUnderBuildDir = "smithyprojections/python/"
```
Comment (crisidev) [inline codegen-server-test/python/build.gradle.kts:19]: I'd like the subfolder to be named `codegen-server-test-python`, but I don't know how to do it. Gradle defaults it to the project folder name, which is `python`.

File: codegen-server/python/build.gradle.kts [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+tasks.jacocoTestReport {
+    reports {
+        xml.isEnabled = false
+        csv.isEnabled = false
+        html.destination = file("$buildDir/reports/jacoco")
+    }
+}
+
+// Always run the jacoco test report after testing.
+tasks["test"].finalizedBy(tasks["jacocoTestReport"])
```
[the remark this answers is not in the document]
Comment (crisidev) [reply codegen-server/python/build.gradle.kts:99]: Oh.. I am going to remove this also from the standard codegen-server project! Thanks.

File: codegen-server/python/build.gradle.kts [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    }
+}
+
+tasks.dokka {
+    outputFormat = "html"
+    outputDirectory = "$buildDir/javadoc"
+}
+
+// Always build documentation
+tasks["build"].finalizedBy(tasks["dokka"])
```
[the remark this answers is not in the document]
Comment (crisidev) [reply codegen-server/python/build.gradle.kts:87]: Oh.. I am going to remove this also from the standard codegen-server project! Thanks.

File: codegen-server/python/build.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+    api("software.amazon.smithy:smithy-codegen-core:$smithyVersion")
+    api("com.moandjiezana.toml:toml4j:0.7.2")
+    implementation(project(":codegen"))
+    implementation(project(":codegen-server"))
+    implementation("software.amazon.smithy:smithy-aws-traits:$smithyVersion")
+    implementation("software.amazon.smithy:smithy-protocol-test-traits:$smithyVersion")
+    implementation("software.amazon.smithy:smithy-waiters:$smithyVersion")
+    runtimeOnly(project(":rust-runtime"))
+    testImplementation("org.junit.jupiter:junit-jupiter:5.6.1")
+    testImplementation("io.kotest:kotest-assertions-core-jvm:$kotestVersion")
```
[the remark this answers is not in the document]
Comment (crisidev) [reply codegen-server/python/build.gradle.kts:40]: Yeah, you are right. I'll do the same for the standard codegen-server project.

File: codegen-server-test/python/build.gradle.kts (hunk 1 above)
Comment (jdisanti) [reply codegen-server-test/python/build.gradle.kts:19]: The Smithy Gradle plugin has a configuration option for [`outputDirectory`](https://awslabs.github.io/smithy/1.0/guides/building-models/gradle-plugin.html). I haven't used it though, so can't say for certain if it'll do what you're wanting.

File: codegen-server/python/build.gradle.kts [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    implementation(project(":codegen"))
+    implementation(project(":codegen-server"))
+    implementation("software.amazon.smithy:smithy-aws-traits:$smithyVersion")
+    implementation("software.amazon.smithy:smithy-protocol-test-traits:$smithyVersion")
+    testImplementation("org.junit.jupiter:junit-jupiter:5.6.1")
+    testImplementation("io.kotest:kotest-assertions-core-jvm:$kotestVersion")
+}
+
+// unlike the client-runtime, software.amazon.smithy.rust.codegen.smithy-kotlin codegen package is
+// not expected to run on Android...we can target 1.8
```
Comment (jdisanti) [inline codegen-server/python/build.gradle.kts:37]: This comment seems out of place. Did it originate from smithy-kotlin or something?

File: codegen-server/python/build.gradle.kts [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+
+publishing {
+    publications {
+        create<MavenPublication>("default") {
+            from(components["java"])
+            artifact(sourcesJar)
+        }
+    }
+    repositories { maven { url = uri("$buildDir/repository") } }
+}
```
Comment (jdisanti) [inline codegen-server/python/build.gradle.kts:82]: Should this get set up later?

File: codegen-server/python/src/main/kotlin/software/amazon/smithy/rust/codegen/server/python/smithy/RustCodegenServerPlugin.kt [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+/** Rust Codegen Plugin
+ *  This is the entrypoint for code generation, triggered by the smithy-build plugin.
+ *  `resources/META-INF.services/software.amazon.smithy.build.SmithyBuildPlugin` refers to this class by name which
+ *  enables the smithy-build plugin to invoke `execute` with all of the Smithy plugin context + models.
+ */
+class RustCodegenServerPlugin : SmithyBuildPlugin {
+    private val logger = Logger.getLogger(javaClass.name)
+
+    override fun getName(): String = "rust-server-codegen"
```
Comment (jdisanti) [inline codegen-server/python/src/main/kotlin/software/amazon/smithy/rust/codegen/server/python/smithy/RustCodegenServerPlugin.kt:34]: Should this have Python somewhere in the name?

File: codegen-server/python/build.gradle.kts (hunk 6 above)
Comment (jdisanti) [reply codegen-server/python/build.gradle.kts:79]: This looks like it's config to publish jars to Maven, but we don't actually publish to Maven yet. I don't think it harms anything though 🤔

Comment (crisidev) [reply codegen-server/python/build.gradle.kts:79]: The same is also present inside `codegen` and `codegen-server` `build.gradle.kt`. Honestly I have just copied it from the `codegen` folder.

File: codegen-server-test/python/build.gradle.kts [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+    commandLine("cargo", "doc", "--no-deps")
+    dependsOn("assemble")
+}
+
+tasks.register<Exec>(Cargo.CLIPPY.toString) {
+    workingDir("$buildDir/$workingDirUnderBuildDir")
+    environment("RUSTFLAGS", defaultRustFlags)
+    commandLine("cargo", "clippy")
+    dependsOn("assemble")
+}
```
[the remark this answers is not in the document]
Comment (crisidev) [reply codegen-server-test/python/build.gradle.kts:103]: If this is not a blocker for this PR, I have created #1378 to track this issue and solve it once I figure out how :)

**Review by 82marbag — APPROVED** [review]

**Review by rcoh — APPROVED** [review]

**Pull request merged**